### PR TITLE
feat: app integrity setup

### PR DESCRIPTION
**Pull request by obinns-dd**
# DCMAW-10308 | App Integrity setup

This pull request completes some setup tasks for implementing app integrity checks using [Firebase App Check](https://firebase.google.com/docs/app-check).

Specifically, we've created a new AppIntegrity Swift Package to contain the logic required to do this. This package contains a new `AppIntegrityProvider` protocol which is able to add integrity assertion headers to a URL request (such as the `/token` request).

The `AppIntegrityProvider` protocol is implemented by a `FirebaseAppIntegrityService` type. This type is currently partially implemented and contains the outline of the full required flow outlines in the [tech design](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/4359061505/Strategic+App+STS+App+Integrity+and+Client+Authentication#Firebase-App-Check).

The current implementation is able to make the `/client-attestation` request, but does not yet:
- provide the Proof-of-Possession key in the request body
- decode the server response
- cache the Client Attestation JWT for reuse

The implementation of `addIntegrityAssertions`, required for `AppIntegrityProvider` conformance, currently attaches mock values for `OAuth-Client-Attestation` and `OAuth-Client-Attestation-PoP`

**The app does not _yet_ import the `AppIntegrity` Swift Package. Therefore this code is currently dormant.**

# Checklist

## Before raising your pull request:
- [x] Commit messages that conform to conventional commit messages
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with a short description about the feature or update
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed

- [ ] ~~Met all accessibility requirements?~~ - N/A : no user facing changes.
   ~- [ ] Checked dynamic type sizes are applied~
    ~- [ ] Checked VoiceOver can navigate your new code~
    ~- [ ] Checked a user can navigate only using a keyboard around your new code~

## Before merging your pull request:
- [x] Ensure that the code coverage and SonarCloud checks have passed
- [x] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [x] Ran the app to ensure that no regressions have been caused by changes during code review.

### CODE REVIEW

File: AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import Foundation
+
+extension URLRequest {
+    static func clientAttestation(token: String) throws -> URLRequest {
+        var components = URLComponents()
+        components.scheme = "https"
+        components.host = "app-integrity-spike.mobile.dev.account.gov.uk"
```
Comment (dylanjames-dd) [inline AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift:7]: Is there a plan to get our environment base urls in here?

Comment (obinns-dd) [reply AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift:7]: Yes, it's gonna be the same format as the `MobilePlatformServices` package uses 🙂

File: AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        components.path = "/client-attestation"
+        components.queryItems = [URLQueryItem(name: "device", value: "ios")]
+
+        let encoder = JSONEncoder()
+
+        var request = URLRequest(url: components.url!)
+        request.httpMethod = "POST"
+        request.httpBody = try encoder.encode(ClientAssertionRequest(jwk: token))
+        request.addValue(token, forHTTPHeaderField: "X-Firebase-AppCheck")
```
Comment (dylanjames-dd) [inline AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift:16]: I'm a little confused as the TD says:
"App calls Mobile Backend /client-attestation endpoint with Firebase App Check token and public app instance key. Mobile Backend validates App Check token claims and signature to validate that Firebase App Check has signed the token and that it is for the One Login app."
We seem to be passing the App Check token twice, once in the header for "X-Firebase-AppCheck" and again in the body for `{"jwk": "token"}`.
I can see on the TD it mentions validating the JWT claims which pertain to Firebase:
image 2024-10-01 at 11 48 48
so not sure where we are including the public key from the app-generated key pair?

Comment (obinns-dd) [reply AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift:16]: Yes, the body is currently wrong, this is WIP 😅 
See here: https://govukverify.atlassian.net/browse/DCMAW-10320

Comment (dylanjames-dd) [reply AppIntegrity/Sources/AppIntegrity/URLRequest+Integrity.swift:11]: I think PoP JWT should be sent to the `/token` endpoint, I thought this was supposed the public key we generate in-app.
Anyway, something we can hash out later 👍

**Review by dylanjames-dd — APPROVED** [review]

**Review by Magdaline-A — APPROVED** [review]

**Pull request merged**